### PR TITLE
Created a dockerfile for local development

### LINKED ISSUE #3: Making a Dockerfile for easy setup
Yeah, currently, the setup process is a painful mess, a dockerfile (or even just a simple build script) would be a good option to make this easier. Even though simple shell scripts would work too, a Dockerfile would be prefered since Docker takes care of everything needed in the background and you can easily start / stop the software.

I will look into this in the future, if someone voluteers, even better :)

### PR DESCRIPTION
**Pull request by FrikFry**
This should resolve #3 

Would appreciate it if someone could run through the README and the Dockerfile and let me know if I'm doing something strange.

### CODE REVIEW

**Comment by Olympic1** [comment]
Looks good and works on Windows, if someone else could test this one more time. More tests are always better then just 1-2.

**Comment by ILM126** [comment]
I can test it later today, when I get home in about 10 hours >.>
If this can wait that long ^_^"

**Comment by phmayo** [comment]
It can wait this long. We are past the rush phase, and can now enjoy the
regular chaos of distributed development.

**Comment by ILM126** [comment]
Alright then, I'll still give it a go then. ^_^
Rushed Dev ftw (first time :P)

**Comment by jonasrosland** [comment]
Nice work @FrikFry, just a suggestion, you can use supervisord (https://docs.docker.com/engine/admin/using_supervisord/) to run multiple applications in one container instead of starting them all on a long RUN line.

**Comment by FrikFry** [comment]
@jonasrosland Thanks for the suggestion. Took a stab at it. Let me know your thoughts please.

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,46 @@
+FROM ubuntu:14.04
+MAINTAINER frikfry@gmail.com # Someone let me know an appropriate email for the project
+
+# Needed to fix pip install of requirements due to strange char encoding issue.
+ENV LC_CTYPE C.UTF-8
+RUN apt-get clean
+RUN apt-get update
+RUN apt-get install -y vim supervisor postgresql postgresql-contrib postgresql-client libpq-dev libffi-dev nodejs node npm python-pip python-dev python3-dev build-essential
```
Comment (Sikian) [inline Dockerfile]: Please consider using node.js' recommended ppa (https://nodejs.org/en/download/package-manager/#debian-and-ubuntu-based-linux-distributions) as the default one may be quite out of date

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+RUN pip install --upgrade pip
+RUN pip install virtualenv
+
+# Make our directories
+RUN mkdir -p /opt/spacedock
+WORKDIR /opt/spacedock
+
+# Install coffee-script
+RUN npm install coffee-script
```
Comment (Sikian) [inline Dockerfile]: Consider using -g. Not a biggy, tho.

**Comment by Sikian** [comment]
Really happy with the pull.
I'll run it tomorrow in a box to double check.

**Comment by Sikian** [comment]
Sorry to keep on pushing this stuff, but I just saw a recommendation I was unaware of: https://docs.docker.com/engine/userguide/eng-image/dockerfile_best-practices/#add-or-copy

I don't think it makes a huge difference, but it might be nice to know :)

**Comment by FrikFry** [comment]
@Sikian Most recent push should resolve the COPY/ADD thing.

I also added a user:development user alongside the previous admin one.

**Comment by Sikian** [comment]
@FrikFry you're a beast!

**Comment by absoutherland** [comment]
Typically speaking, you really only want 1 process per container. If 1 service flakes, do you restart all of the services?

Maybe using something like docker-compose to orchestrate the cluster? 

All in all, this definitely goes hard. I'll take a crack at orchestration later on this evening.

**Comment by FrikFry** [comment]
@absoutherland I'm not very familiar with docker-compose but it looks interesting. I think it might be worth it just for getting essentially a config file for docker-run so that people don't have to worry about copying that big nasty line in my README around.

However, I am concerned about splitting up all of the services into their own containers. It seems like the project as it currently stands makes lots of assumptions that everything is running in localhost. For example, the default config file assumes localhost for the postgres db. Additionally, redis is currently hardcoded as localhost in the source here:
https://github.com/KSP-SpaceDock/SpaceDock/blob/master/KerbalStuff/celery.py#L6

I'm not very familiar with how docker-compose handles networking like that, but I expect it'll probably want a different host than ```localhost``` for redis, or any other service. If so, it seems like we'd have to manage a config.ini.docker alongside the config.ini.example and/or make modifications to the config.ini file on the container after it's copied, and I'm not sure how maintainable that is in the long run. I think the ideal workflow would be to run one or two commands and have a working development server, and not need to worry about modifying a config.ini file manually.

I don't think it's a big deal to expect people to have to occasionally restart services and/or the container when the goal is enabling speedy local development. If my concerns are justified above, maybe we can revisit splitting all of the dependent services into their own containers after we fix up the whole config.ini area a bit.

@ThomasKerman You originally called for this effort. Do you want to weigh in on this?

**Comment by ThomasKerman** [comment]
Well, I always follow the motto "never change a running system" and as your Dockerfile works perfectly, I would just merge it, and don't care about how well written it is in the background - as long as it works for the end user. :smiley:

**Comment by phmayo** [comment]
Well, we do have to consider maintenance on our end, as well.

@FrikFry I would split the services up (1 per container), and then we can see what breaks, and needs to be configured within its container and the whole stack. Just don't use the Docker containers in production just yet.

**Comment by FrikFry** [comment]
I did some more reading on docker-compose and it looks like I was probably making mountains out of molehills with regards to the difficulty of setting up the separate containers.

One question I do have is that it probably makes sense to tailor config.ini.example and alembic.ini.example for use by Docker with the expectation that if you want to run SpaceDock in another manner, you'd probably want to modify those configs anyway. Anyone have an objection to that?

**Comment by ThomasKerman** [comment]
Do we maybe want to merge this and do the docker composing in a new PR? That would improve organization, and VITAS could set up the container on his server already, which would improve various things.

**Comment by FrikFry** [comment]
@ThomasKerman I agree with merging it now, but I don't know how helpful it will be in its current state to VITAS, as I don't seem to have email working. I'll push another version with redis and celery services starting and perhaps someone can debug the mail issue? At the moment I was trying to use a gmail account but that doesn't seem to be working.

**Comment by FrikFry** [comment]
@ThomasKerman @phmayo I feel pretty good about just merging this now and handling docker-compose with appropriate splitting up of services/container in a separate PR. Any dissent?

**Comment by ThomasKerman** [comment]
LGTM

**Pull request merged**